File: app/src/main/java/com/kaciula/archiman/domain/usecases/InitColdStart.java
package com.kaciula.archiman.domain.usecases;

import com.google.auto.value.AutoValue;
import com.kaciula.archiman.domain.boundary.data.AppInfoRepository;
import com.kaciula.archiman.domain.util.UseCase;
import io.reactivex.Completable;
import io.reactivex.Observable;
import timber.log.Timber;

public class InitColdStart
    extends UseCase<InitColdStart.RequestModel, InitColdStart.ResponseModel> {

  private final AppInfoRepository appInfoRepository;

  public InitColdStart(AppInfoRepository appInfoRepository) {
    this.appInfoRepository = appInfoRepository;
  }

  @Override
  public Observable<ResponseModel> execute(final RequestModel requestModel) {
    return Completable.fromAction(() -> {
      Timber.d("Initialize every cold start");
      int currentVersionCode = requestModel.currentVersionCode();
      if (appInfoRepository.isFirstTime()) {
        Timber.d("First time running the app");
        appInfoRepository.saveFirstTime(false);
        appInfoRepository.saveVersionCode(currentVersionCode);
      } else {
        if (appInfoRepository.getVersionCode() < currentVersionCode) {
          Timber.d("Old version code %d is replaced with new version code %d",
              appInfoRepository.getVersionCode(), currentVersionCode);
          appInfoRepository.saveVersionCode(currentVersionCode);
        } else {
          Timber.d("Just a basic cold start");
        }
      }
    }).toObservable();
  }

  @AutoValue
  public abstract static class RequestModel implements UseCase.RequestModel {
    public abstract int currentVersionCode();

    public static RequestModel create(int currentVersionCode) {
      return new AutoValue_InitColdStart_RequestModel(currentVersionCode);
    }
  }


  @AutoValue
  abstract static class ResponseModel implements UseCase.ResponseModel {
    public static ResponseModel create() {
      return new AutoValue_InitColdStart_ResponseModel();
    }
  }
}

File: app/src/main/java/com/kaciula/archiman/domain/usecases/InitColdStart.kt
package com.kaciula.archiman.domain.usecases

import com.kaciula.archiman.boundary.AppRepository
import com.kaciula.archiman.domain.usecases.util.CompletableUseCase
import io.reactivex.Completable
import timber.log.Timber

class InitColdStart(private val appRepository: AppRepository) :
    CompletableUseCase<InitColdStart.RequestModel>() {

    override fun execute(requestModel: RequestModel): Completable {
        return Completable.fromAction {
            Timber.i("Initialize every cold start")
            val currentVersionCode = requestModel.currentVersionCode
            val app = appRepository.get()
            if (app.shouldBumpVersion(currentVersionCode)) {
                app.bumpVersion(currentVersionCode)
                appRepository.save(app)
            } else {
                Timber.i("Just a basic cold start")
            }
        }
    }

    data class RequestModel(val currentVersionCode: Int)
}
